### ISSUE TITLE
The editor key binding does not follow the physical keyboard layout to move the viewport camera

### ISSUE THREAD

**Issue by Dragnalith** [OurMachinery/themachinery-public#870]
To move the camera in the editor viewport I have to press W, A, S, D, but depending on keyboard layout (BÉPO, DVORAK, AZERTY) W, A, S, D are not convenient (it is not map similar to UP, LEFT, DOWN, RIGHT).

For instance in BÉPO I would like camera movement to be É, A, U, I keys.

I think it is nice it the mapping for those key is mapping according to the keyboard physical layout and not the language layout.

**Comment by niklas-ourmachinery** [comment]
You can configure what keys to use with:

**File > Settings > Keyboard Shortcuts > Scene**

This way you can set up keys that work for your keyboard layout.

It still depends on your keyboard layout though, so if you switch from QWERTY to DVORAK (say), you will have to go into the settings again and update the keys.

We could base it on physical keys (i.e. scan codes) instead, but I'm not sure if that is a good idea. AFAIK, most programs that use shortcuts use key names rather than scan codes.

But maybe the first time you run The Machinery, we should see what kind of keyboard layout you have and set suitable default keys based on that. It's a bit annoying to have to go in and change the settings the first thing you do.

**Comment by Dragnalith** [comment]
From my experience of using a very different keyboard layout I distinguish  2 kind of keybinding:
- A/ The one for command (copy Ctrl+C, select Ctrl-A, save Ctrl-S, build Ctrl-B)
- B/ The one for realtime interaction (move W, A, S, D, crouch C, action F)

The A/ type does not really matter if it follows the keyboard physical layout, there are so much command that some cannot be ergonomic. 

For the B/ type it is important to follow the physical layout and have a good ergonomy. Most of modern game have physical keybinding (physical binding does not even change if you modify the keyword layout while game is running)

In the case of TM Editor I think most of command should be A/ as it is today, except for the viewport movement which should follow the physical layout.

(since viewport movement is only enabled when mouse clicking it, it should not conflict with command key binding)

**Comment by niklas-ourmachinery** [comment]
Yes, that makes a lot of sense... however it would mean that in our keybinding editor we need to distinguish between physical/virtual keybindings which I'm worried might be a bit confusing too people...

It seems to me that using virtual keybindings for WASD navigation works well except for two issues:

1. Initially, the setup might be bad for people who are not on QWERTY keyboards, meaning they can't really navigate until they've gone in and changed the settings.
2. For people who switch frequently between different keyboard layouts, it might be annoying, because the keys they set up for one layout might not work for others.

(1) it seems we could fix by just using better defaults, based on the user's keyboard layout when they first start the program.

(2) I'm not sure how big of a problem this actually is... since we support primary and secondary keybindings, only users who switch between more than two different keyboard layouts would be affected.

So it seems we could do a pretty good solution to this without introducing the concept of physical/virtual bindings... still might be worth to have that concept for games to use though...

**Comment by Dragnalith** [comment]
I feel the scenario (1) would be good enough.

One issue came to my mind, is it is possible for me to boot TM with the
wrong layout (I am often switching between QWERTY, DVORAK, and Japanese
IME). Unless if TM is sensitive to logical keyboard layout I don't see
solution for this problem.

(Also, I am used to Ctrl-C, Ctrl-V, Ctrl-X, Ctrl-A, ... whatever the
layout. I only wish for "in-game" control (i.e W, A, S, D) to be
different).

Regarding game, my experience as a gamer is to be very grateful when the
game use physical layout instead of default one. (I would say 50% of big
company game do that)
